1.1  Answer to these questions with at least one example for each .


In [ ]:
#What is the difference between static and dynamic variables in Python?

Difference Between Static and Dynamic Variables in Python

**Static Variables:** Also known as class variables, static variables are shared among all instances of a class. They are defined at the class level and can be accessed using the class name or an instance of the class.

**Dynamic Variables:** Also known as instance variables, dynamic variables are unique to each instance of a class. They are defined within methods (typically __init__) and are prefixed with self.

In [ ]:
class MyClass:
    # Static variable (class variable)
    static_var = "I am a static variable"

# Creating instances
obj1 = MyClass()
obj2 = MyClass()

# Accessing static variable through class
print(MyClass.static_var)  # Output: I am a static variable

# Accessing static variable through instances
print(obj1.static_var)     # Output: I am a static variable
print(obj2.static_var)     # Output: I am a static variable

# Modifying static variable through class
MyClass.static_var = "Changed static variable"
print(obj1.static_var)     # Output: Changed static variable
print(obj2.static_var)     # Output: Changed static variable


In [ ]:
class MyClass:
    def __init__(self, dynamic_var):
        # Dynamic variable (instance variable)
        self.dynamic_var = dynamic_var

# Creating instances with different dynamic variables
obj1 = MyClass("I am a dynamic variable in obj1")
obj2 = MyClass("I am a dynamic variable in obj2")

# Accessing dynamic variables through instances
print(obj1.dynamic_var)    # Output: I am a dynamic variable in obj1
print(obj2.dynamic_var)    # Output: I am a dynamic variable in obj2

# Modifying dynamic variable of obj1
obj1.dynamic_var = "Changed dynamic variable in obj1"
print(obj1.dynamic_var)    # Output: Changed dynamic variable in obj1
print(obj2.dynamic_var)    # Output: I am a dynamic variable in obj2


In [ ]:
#• Explain the purpose of "pop","popitem","clear()" in a dictionary with suitable examples.

In Python, dictionaries have several methods to manipulate the data they contain. The methods pop(), popitem(), and clear() are used to remove elements from a dictionary. Here is an explanation of each method along with suitable examples.

pop()
Purpose: The pop() method removes the specified key from the dictionary and returns its value. If the key is not found, it raises a KeyError unless a default value is provided.

Example:

In [ ]:
my_dict = {'a': 1, 'b': 2, 'c': 3}
# Remove and return the value for the key 'b'
value = my_dict.pop('b')
print(value)  # Output: 2
print(my_dict)  # Output: {'a': 1, 'c': 3}

# Remove a key that doesn't exist with a default value
value = my_dict.pop('d', 'Key not found')
print(value)  # Output: Key not found


popitem()
Purpose: The popitem() method removes and returns the last (key, value) pair inserted into the dictionary. This method is useful for implementing LIFO (last-in, first-out) order. If the dictionary is empty, it raises a KeyError.

Example:

In [ ]:
my_dict = {'a': 1, 'b': 2, 'c': 3}
# Remove and return the last inserted item
item = my_dict.popitem()
print(item)  # Output: ('c', 3)
print(my_dict)  # Output: {'a': 1, 'b': 2}

# Remove another item
item = my_dict.popitem()
print(item)  # Output: ('b', 2)
print(my_dict)  # Output: {'a': 1}


clear()
Purpose: The clear() method removes all items from the dictionary, resulting in an empty dictionary.

Example:

In [ ]:
my_dict = {'a': 1, 'b': 2, 'c': 3}
# Clear all items from the dictionary
my_dict.clear()
print(my_dict)  # Output: {}


pop(key[, default]): Removes and returns the value for the specified key. Raises KeyError if the key is not found and no default is provided.

popitem(): Removes and returns the last inserted (key, value) pair. Raises KeyError if the dictionary is empty.

clear(): Removes all items from the dictionary, leaving it empty.

In [ ]:
#What do you mean by FrozenSet? Explain it with suitable examples.

A FrozenSet in Python is an immutable collection of unique elements, similar to a set, but once created, its elements cannot be changed. Here's a brief example:
In this example, frozen_set is a frozenset containing unique elements [1, 2, 3, 4]. Once created, you cannot modify its contents (e.g., adding or removing elements), making it useful for scenarios where immutability is desired.

In [ ]:
# Creating a FrozenSet
frozen_set = frozenset([1, 2, 3, 4, 4])

# Trying to add elements (will raise an error)
# frozen_set.add(5)

print(frozen_set)  # Output: frozenset({1, 2, 3, 4})


In [ ]:
#• Differentiate between mutable and immutable data types in Python and give examples of mutable and immutable data types.


Mutable data types can be changed after creation (e.g., lists, dictionaries); immutable data types cannot be changed once created (e.g., tuples, strings).
Examples:

Mutable: list (e.g., [1, 2, 3]), dict (e.g., {'a': 1, 'b': 2})
Immutable: tuple (e.g., (1, 2, 3)), str (e.g., 'hello')







In [ ]:
#• What is __init__?Explain with an example.


__init__ is a special method in Python classes used to initialize objects when they are created, allowing you to set initial values for attributes.

In this example, __init__ initializes each instance of Car with make and model attributes ("Toyota" and "Camry" for my_car).

In [ ]:
class Car:
    def __init__(self, make, model):
        self.make = make
        self.model = model

# Creating an instance of Car
my_car = Car("Toyota", "Camry")


In [ ]:
#• What is docstring in Python?Explain with an example.

A docstring in Python is a string literal that appears right after the definition of a function, method, class, or module. It is used to describe what the object does. Docstrings are accessed using the __doc__ attribute.

In this example, """This function greets the person with the given name.""" is the docstring for the greet function. It provides a concise description of what the function does.

In [ ]:
def greet(name):
    """
    This function greets the person with the given name.
    """
    print(f"Hello, {name}!")

# Accessing the docstring
print(greet.__doc__)


In [ ]:
#• What are unit tests in Python?


Unit tests in Python are automated tests written to verify that specific parts of code (like functions or classes) produce expected results under different conditions, ensuring code correctness and reliability.
They are an integral part of Test-Driven Development (TDD) and Continuous Integration (CI) workflows, contributing to software quality by catching bugs early in the development process.

In [ ]:
#• What is break, continue and pass in Python?



break: Terminates the current loop statement (e.g., for or while) and transfers execution to the statement immediately following the loop.

continue: Skips the current iteration of a loop (e.g., for or while) and continues with the next iteration.

pass: Acts as a placeholder in Python syntax, doing nothing when executed; used as a filler for code blocks where no action is needed syntactically.

In [ ]:
#• What is the use of self in Python?


In Python, self is used inside methods to refer to the instance of the class itself, allowing methods to access and manipulate instance variables and other methods within the same class. It is a convention in Python but is not a reserved keyword.

In this example:

self.make and self.model are instance variables initialized in the __init__ method using self.
display_info is a method that uses self to access and print the instance variables make and model of the Car object (my_car).

In [ ]:
class Car:
    def __init__(self, make, model):
        self.make = make
        self.model = model

    def display_info(self):
        print(f"Car make: {self.make}, model: {self.model}")

# Creating an instance of Car
my_car = Car("Toyota", "Camry")

# Calling a method using the instance
my_car.display_info()


In [ ]:
#• What are global, protected and private attributes in Python?


Global Attributes: These are accessible throughout the entire program, regardless of scope.

Protected Attributes: By convention, attributes prefixed with a single underscore _ are intended to be protected, indicating that they should not be accessed directly outside the class or its subclasses unless necessary.

Private Attributes: Attributes prefixed with double underscore __ are considered private, meaning they cannot be accessed directly from outside the class. Python name-mangles these attributes to avoid accidental overriding in subclasses.

In [ ]:
global_var = 10

def func():
    print(global_var)  # Can access global_var

func()


In [ ]:
class Car:
    def __init__(self, make, model):
        self._make = make
        self._model = model

    def display_info(self):
        print(f"Car make: {self._make}, model: {self._model}")

my_car = Car("Toyota", "Camry")
print(my_car._make)  # Accessing protected attribute (not enforced by Python, but by convention)


In [ ]:
class Car:
    def __init__(self, make, model):
        self.__make = make
        self.__model = model

    def display_info(self):
        print(f"Car make: {self.__make}, model: {self.__model}")

my_car = Car("Toyota", "Camry")
# print(my_car.__make)  # This would raise an AttributeError because __make is private


In [ ]:
#• What are modules and packages in Python?


Modules are files containing Python code that define functions, classes, and variables. They can be imported into other Python scripts to reuse code.

Packages are directories containing multiple modules and an __init__.py file. They provide a hierarchical structure to organize and import modules.

In [ ]:
# Example module named mymodule.py
def greet(name):
    print(f"Hello, {name}!")

# Using the module in another script
import mymodule

mymodule.greet("Alice")


In [ ]:
mypackage/
├── __init__.py
├── module1.py
└── module2.py

# Using modules from a package
from mypackage import module1, module2

module1.function1()
module2.function2()

In [ ]:
#• What are lists and tuples? What is the key difference between the two?


Lists are mutable sequences that can contain elements of different types and can be modified after creation using methods like append() or slicing operations.

Tuples are immutable sequences that can contain elements of different types but cannot be modified after creation. They are defined using parentheses.

Key Difference: Lists are mutable, meaning they can be changed after creation, while tuples are immutable and cannot be modified once created.

In [ ]:
# Example of a list
my_list = [1, 2, 3, 'apple', 'banana']
my_list.append(4)
print(my_list)  # Output: [1, 2, 3, 'apple', 'banana', 4]


In [ ]:
# Example of a tuple
my_tuple = (1, 2, 3, 'apple', 'banana')
# my_tuple.append(4)  # This would raise an AttributeError because tuples are immutable
print(my_tuple)  # Output: (1, 2, 3, 'apple', 'banana')


In [ ]:
#• What is an Interpreted language & dynamically typed language?Write 5 differences between them.


Interpreted Language:

Executes code directly without compilation.
Typically slower than compiled languages.
Examples include Python, JavaScript, and Ruby.
Dynamically Typed Language:

Variable types are determined at runtime.
Allows flexibility but may lead to runtime errors.
Examples include Python, JavaScript, and Ruby.
Differences:

Execution Method:

Interpreted languages execute code line-by-line without prior compilation.
Dynamically typed languages determine variable types at runtime.
Speed:

Interpreted languages tend to be slower because they translate code during execution.
Dynamically typed languages may involve runtime type checks, impacting performance.
Error Detection:

Interpreted languages often catch errors during runtime.
Dynamically typed languages may encounter type-related errors during execution.
Flexibility vs. Safety:

Interpreted languages provide flexibility with immediate feedback but less safety.
Dynamically typed languages offer flexibility but may result in unexpected runtime errors.

In [ ]:
# Python is both an interpreted and dynamically typed language:
# Interpreted and dynamically typed example in Python
x = 10
x = "Hello, World!"  # Valid in a dynamically typed language like Python
print(x)


In [ ]:
#• What are Dict and List comprehensions?


List Comprehension: A concise way to create lists by iterating over an iterable and applying an expression to each element.

Dict Comprehension: Similar to list comprehension but creates dictionaries instead of lists, allowing you to generate dictionaries from iterables.

Key Differences:

List comprehensions create lists by iterating over an iterable.
Dict comprehensions create dictionaries by iterating over an iterable and specifying key-value pairs.
Both comprehensions are concise and efficient ways to generate collections in Python.

In [ ]:
# Example of list comprehension
squares = [x**2 for x in range(1, 6)]
print(squares)  # Output: [1, 4, 9, 16, 25]


In [ ]:
# Example of dict comprehension
squares_dict = {x: x**2 for x in range(1, 6)}
print(squares_dict)  # Output: {1: 1, 2: 4, 3: 9, 4: 16, 5: 25}


In [ ]:
#• What are decorators in Python? Explain it with an example.Write down its use cases.


Decorators in Python are a way to modify or enhance the behavior of functions or methods. They allow you to wrap another function in order to extend its behavior without permanently modifying it.


Use Cases
Logging: Adding logging to functions.
Authorization: Checking if a user has permissions to execute a function.
Caching: Storing results of expensive function calls and reusing cached results.
Validation: Validating inputs to functions.
Measurement: Measuring the execution time of functions.

In this example, my_decorator is a decorator that adds extra behavior before and after the say_hello function.


In [ ]:
def my_decorator(func):
    def wrapper():
        print("Something is happening before the function is called.")
        func()
        print("Something is happening after the function is called.")
    return wrapper

@my_decorator
def say_hello():
    print("Hello!")

say_hello()


In [ ]:
#• How is memory managed in Python?


Memory management in Python is handled by the Python memory manager, which includes an in-built garbage collector for automatic recycling of unused memory.

Key Points
Reference Counting: Python uses reference counting to keep track of the number of references to each object in memory.
Garbage Collection: When the reference count drops to zero, the memory occupied by the object is reclaimed. Python also has a cyclic garbage collector to detect and collect cyclic references.

In this example, the list [1, 2, 3] is automatically removed from memory when there are no more references to it.

In [ ]:
a = [1, 2, 3]
b = a  # Both 'a' and 'b' reference the same list
del a  # The reference count for the list is still 1 because 'b' still references it
del b  # The reference count drops to 0, so the memory is reclaimed by the garbage collector


In [ ]:
#• What is lambda in Python? Why is it used?


A lambda in Python is an anonymous (unnamed) function defined using the lambda keyword. It can have any number of arguments but only one expression.

Why Use Lambda?
Shorter Syntax: For defining small, throwaway functions without using def.
Inline Functions: Used in places where functions are required but defining a full function would be overkill (e.g., within map(), filter(), etc.).

In this example, lambda x: x**2 defines an anonymous function that squares its input. This function is used within the map() function to apply to each element in the numbers list.

In [ ]:
# Using lambda with map to square numbers in a list
numbers = [1, 2, 3, 4]
squared = list(map(lambda x: x**2, numbers))
print(squared)  # Output: [1, 4, 9, 16]


In [ ]:
#• Explain split() and join() functions in Python?


split() Function
Purpose: Splits a string into a list of substrings based on a specified delimiter (default is whitespace).

join() Function
Purpose: Joins elements of a list (or any iterable) into a single string with a specified delimiter.

split() breaks a string into parts, and join() combines parts into a string.

In [ ]:
text = "Hello World"
words = text.split()
print(words)  # Output: ['Hello', 'World']


In [ ]:
words = ['Hello', 'World']
text = " ".join(words)
print(text)  # Output: "Hello World"


In [ ]:
#• What are iterators, iterable & generators in Python?


Iterators
Definition: An object that allows traversal through all the elements of a collection, one at a time.
Methods: Implements __iter__() and __next__() methods.

Iterables
Definition: An object that can return an iterator, allowing it to be looped over.
Methods: Implements __iter__() method.

Generators
Definition: A special type of iterator created using a function that yields values one at a time.
Methods: Uses yield keyword to produce a series of values.

In summary, iterables are collections that can be looped over, iterators are objects used to traverse these collections, and generators are a convenient way to create iterators using functions and the yield keyword.

In [ ]:
numbers = [1, 2, 3]
iterator = iter(numbers)
print(next(iterator))  # Output: 1
print(next(iterator))  # Output: 2


In [ ]:
numbers = [1, 2, 3]
for number in numbers:  # `numbers` is an iterable
    print(number)


In [ ]:
def count_up_to(max):
    count = 1
    while count <= max:
        yield count
        count += 1

generator = count_up_to(3)
print(next(generator))  # Output: 1
print(next(generator))  # Output: 2


In [ ]:
#• What is the difference between xrange and range in Python?


In Python 2, xrange and range are used to generate sequences of numbers, but they function differently.

range (Python 2 and 3)
Type: Returns a list (Python 2) or a range object (Python 3).
Memory Usage: Stores all numbers in memory (Python 2).
xrange (Python 2 only)
Type: Returns an xrange object, which generates numbers on demand (lazy evaluation).
Memory Usage: More memory efficient for large ranges.

In Python 3
xrange is removed.
range behaves like xrange, returning an immutable sequence object that generates numbers on demand.

In summary, range in Python 3 combines the functionalities of both range and xrange from Python 2, being both memory efficient and versatile.

In [ ]:
# range example (Python 2)
r = range(1, 5)
print(r)  # Output: [1, 2, 3, 4]

# xrange example (Python 2)
xr = xrange(1, 5)
print(list(xr))  # Output: [1, 2, 3, 4]

# range example (Python 3)
r = range(1, 5)
print(list(r))  # Output: [1, 2, 3, 4]


In [ ]:
# Pillars of Oops.


Encapsulation

Definition: Bundling of data (attributes) and methods (functions) that operate on the data into a single unit (class), and restricting access to some of the object's components.

Inheritance

Definition: Mechanism where a new class inherits attributes and methods from an existing class.

Polymorphism

Definition: Ability to present the same interface for different data types or classes.

Abstraction

Definition: Hiding complex implementation details and showing only the necessary features of an object.

In summary, the four pillars of OOP—encapsulation, inheritance, polymorphism, and abstraction—provide a framework for building modular, reusable, and maintainable code.

In [ ]:
#Encapsulation
class Person:
    def __init__(self, name, age):
        self.name = name
        self.__age = age  # Private attribute

    def get_age(self):
        return self.__age

p = Person("Alice", 30)
print(p.name)       # Output: Alice
print(p.get_age())  # Output: 30

#Inheritance
class Animal:
    def speak(self):
        return "Some sound"

class Dog(Animal):
    def speak(self):
        return "Bark"

d = Dog()
print(d.speak())  # Output: Bark

#polymorphism
class Cat:
    def speak(self):
        return "Meow"

class Dog:
    def speak(self):
        return "Bark"

animals = [Cat(), Dog()]
for animal in animals:
    print(animal.speak())  # Output: Meow Bark

#abstraction
from abc import ABC, abstractmethod

class Shape(ABC):
    @abstractmethod
    def area(self):
        pass

class Rectangle(Shape):
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height

r = Rectangle(4, 5)
print(r.area())  # Output: 20


In [ ]:
#• How will you check if a class is a child of another class?


To check if a class is a child (subclass) of another class in Python, you can use the issubclass() function.

In this example, issubclass(Child, Parent) returns True because Child is a subclass of Parent.

In [ ]:
class Parent:
    pass

class Child(Parent):
    pass

print(issubclass(Child, Parent))  # Output: True
print(issubclass(Parent, Child))  # Output: False


In [ ]:
#• How does inheritance work in python? Explain all types of inheritance with an example.


In [ ]:
#Inheritance in Python allows a class to inherit attributes and methods from another class, promoting code reuse and hierarchical class organization. The class being inherited from is called the parent or superclass, and the class that inherits is called the child or subclass.

#Types of Inheritance
#Single Inheritance

#A subclass inherits from one superclass.Example:

class Animal:
    def speak(self):
        return "Some sound"

class Dog(Animal):
    def speak(self):
        return "Bark"

d = Dog()
print(d.speak())  # Output: Bark

#Multiple Inheritance

#A subclass inherits from more than one superclass.Example:

class Walker:
    def walk(self):
        return "Walking"

class Swimmer:
    def swim(self):
        return "Swimming"

class Amphibian(Walker, Swimmer):
    pass

a = Amphibian()
print(a.walk())  # Output: Walking
print(a.swim())  # Output: Swimming

#Multilevel Inheritance

#A subclass inherits from a class which itself is a subclass of another class.Example:

class Animal:
    def speak(self):
        return "Some sound"

class Mammal(Animal):
    def has_fur(self):
        return True

class Dog(Mammal):
    def speak(self):
        return "Bark"

d = Dog()
print(d.speak())  # Output: Bark
print(d.has_fur())  # Output: True

#Hierarchical Inheritance

#Multiple subclasses inherit from a single superclass.Example:

class Animal:
    def speak(self):
        return "Some sound"

class Dog(Animal):
    def speak(self):
        return "Bark"

class Cat(Animal):
    def speak(self):
        return "Meow"

d = Dog()
c = Cat()
print(d.speak())  # Output: Bark
print(c.speak())  # Output: Meow
Hybrid Inheritance

#A combination of two or more types of inheritance.Example:

class Animal:
    def speak(self):
        return "Some sound"

class Mammal(Animal):
    def has_fur(self):
        return True

class Bird(Animal):
    def has_feathers(self):
        return True

class Bat(Mammal, Bird):
    def can_fly(self):
        return True

b = Bat()
print(b.speak())      # Output: Some sound
print(b.has_fur())    # Output: True
print(b.has_feathers())  # Output: True
print(b.can_fly())    # Output: True

#In summary, inheritance in Python provides a way to create hierarchical class structures, promoting code reuse and organization.


In [ ]:
#• What is encapsulation? Explain it with an example.


Encapsulation in Python is the concept of bundling data (attributes) and methods (functions that operate on the data) together within a class, and controlling access to that data through methods. It helps in hiding the internal state of an object and preventing external code from directly accessing it, ensuring data integrity and security.

In this example:

make, model, and year are public attributes of the Car class.
__mileage is a private attribute (note the double underscores __), which can only be accessed or modified within the class itself.
The drive() method updates the mileage, encapsulating the logic of how mileage is managed.
The get_mileage() method provides controlled access to the private __mileage attribute, encapsulating its retrieval.
Encapsulation helps in maintaining clean and manageable code by hiding internal implementation details and exposing only necessary interfaces to interact with objects.

In [ ]:
class Car:
    def __init__(self, make, model, year):
        self.make = make       # Public attribute
        self.model = model     # Public attribute
        self.year = year       # Public attribute
        self.__mileage = 0    # Private attribute

    def drive(self, miles):
        """Simulate driving and update mileage."""
        self.__mileage += miles

    def get_mileage(self):
        """Return the current mileage."""
        return self.__mileage

# Creating an instance of Car
my_car = Car("Toyota", "Camry", 2020)

# Accessing public attributes directly
print(my_car.make)   # Output: Toyota
print(my_car.model)  # Output: Camry
print(my_car.year)   # Output: 2020

# Trying to access private attribute directly (will result in AttributeError)
# print(my_car.__mileage)

# Driving the car
my_car.drive(100)
my_car.drive(50)

# Accessing private attribute through a method
print(my_car.get_mileage())  # Output: 150


In [ ]:
#• What is polymorphism? Explain it with an example.

Polymorphism in Python refers to the ability of different objects to respond to the same method or function in different ways. It allows objects of different classes to be treated as objects of a common superclass, providing flexibility and extensibility in code.

In this example:

The Animal class defines a method speak() which is overridden in its subclasses Dog, Cat, and Cow.
The make_sound() function takes an object of type Animal (or any subclass of Animal) as an argument and calls its speak() method.
Despite each subclass implementing speak() differently (Woof!, Meow!, Moo!), the make_sound() function can treat all these objects uniformly and call their speak() method, demonstrating polymorphic behavior.
Polymorphism allows for more flexible and modular code by enabling objects of different types to be used interchangeably in contexts where their superclass is expected, enhancing code reusability and readability.

In [ ]:
class Animal:
    def speak(self):
        pass

class Dog(Animal):
    def speak(self):
        return "Woof!"

class Cat(Animal):
    def speak(self):
        return "Meow!"

class Cow(Animal):
    def speak(self):
        return "Moo!"

# Function demonstrating polymorphism
def make_sound(animal):
    return animal.speak()

# Creating instances of different classes
dog = Dog()
cat = Cat()
cow = Cow()

# Calling the function with different objects
print(make_sound(dog))  # Output: Woof!
print(make_sound(cat))  # Output: Meow!
print(make_sound(cow))  # Output: Moo!


In [ ]:
#1. 2. Which of the following identifier names are invalid and why?
#a) Serial_no. b) 1st Room c) Hundred$ d) Total Marks e) total-Marks f) Total Marks g) True h) _Percentag


The invalid identifiers are:

1st Room (contains a space and starts with a digit)

Total Marks (contains a space)

total-Marks (contains a hyphen)

True (reserved keyword)

Total Marks (contains a space)

In [ ]:
# 1.3.
#name = ["Mohan", "dash", "karam", "chandra","gandhi","Bapu"]
#do the following operations in this list;
#a) add an element "freedom_fighter" in this list at the Oth index.
#b) find the output of the following, and explain how?
#name = ["freedomFighter", "Bapuji", "Mohan" "dash", "karam", "chandra", "gandhi"]
#Length1=len((name [-len (name) +1:-1:2])) Length2=len((name [-Len (name)+1:-1])) print(length1+Length2)


In [ ]:
name = ["Mohan", "dash", "karam", "chandra", "gandhi", "Bapu"]
name.insert(0, "freedom_fighter")
print(name)


['freedom_fighter', 'Mohan', 'dash', 'karam', 'chandra', 'gandhi', 'Bapu']


In [ ]:
name = ["freedomFighter", "Bapuji", "Mohan" "dash", "karam", "chandra", "gandhi"]
Length1 = len(name[-len(name) + 1 : -1 : 2])
Length2 = len(name[-len(name) + 1 : -1])
print(Length1 + Length2)


6


In [ ]:
#c) add two more elements in the name ["Netaji","Bose"] at the end of the list.
#d) what will be the value of temp:
#name = ["Bapuji", "dash", "karam", "chandra","gandi", "Mohan"]
#temp=name[-1]
#name[-1]=name[0] name[0]=temp
#print(name)

In [ ]:
name = ["freedom_fighter", "Mohan", "dash", "karam", "chandra", "gandhi", "Bapu"]
name.extend(["Netaji", "Bose"])
print(name)


['freedom_fighter', 'Mohan', 'dash', 'karam', 'chandra', 'gandhi', 'Bapu', 'Netaji', 'Bose']


In [ ]:
name = ["Bapuji", "dash", "karam", "chandra", "gandi", "Mohan"]
temp = name[-1]
name[-1] = name[0]
name[0] = temp
print(name)


['Mohan', 'dash', 'karam', 'chandra', 'gandi', 'Bapuji']


In [ ]:
#Question 1.4.Find the output of the following.
#animal = ['Human','cat','mat,'cat','rat','Human', 'Lion']
#print(animal.count('Human'))
#print(animal.index('rat'))
#print(len(animal))


In [ ]:
animal = ['Human', 'cat', 'mat', 'cat', 'rat', 'Human', 'Lion']
print(animal.count('Human'))
print(animal.index('rat'))
print(len(animal))


2
4
7


In [ ]:
#Question 1.5. tuple1=(10,20,"Apple",3.4,'a', ["master", "ji"],("sita","geeta",22),[{"roll_no":1}, {"name":"Navneet"}])
#a) print(len(tuple1))
#b) print(tuple1[-1][-1]["name"])
#c) fetch the value of roll_no from this tuple.
#d) print(tuple1[-3][1])
#e)fetch the element "22" from this tuple.


In [ ]:
tuple1 = (10, 20, "Apple", 3.4, 'a', ["master", "ji"], ("sita", "geeta", 22), [{"roll_no": 1}, {"name": "Navneet"}])
print(len(tuple1))
print(tuple1[-1][-1]["name"])
print(tuple1[-1][0]["roll_no"])
print(tuple1[-3][1])
print(tuple1[-2][2])




8
Navneet
1
ji
22


In [ ]:
#1.6. Write a program to display the appropriate message as per the color of signal(RED-Stop/Yellow-Stay/ Green-Go) at the road crossing.

In [ ]:
def traffic_signal(signal_color):
    if signal_color.lower() == 'red':
        return "Stop"
    elif signal_color.lower() == 'yellow':
        return "Stay"
    elif signal_color.lower() == 'green':
        return "Go"
    else:
        return "Invalid color"

# Test the function
signal_color = input("Enter the color of the signal (RED/Yellow/Green): ")
message = traffic_signal(signal_color)
print(message)


Enter the color of the signal (RED/Yellow/Green): yellow
Stay


In [ ]:
#1.7. Write a program to create a simple calculator performing only four basic operations (+,-,/,*).

In [ ]:
def add(x, y):
    return x + y

def subtract(x, y):
    return x - y

def multiply(x, y):
    return x * y

def divide(x, y):
    if y == 0:
        return "Error! Division by zero."
    else:
        return x / y

def calculator():
    print("Select operation:")
    print("1. Add")
    print("2. Subtract")
    print("3. Multiply")
    print("4. Divide")

    choice = input("Enter choice(1/2/3/4): ")

    if choice in ['1', '2', '3', '4']:
        num1 = float(input("Enter first number: "))
        num2 = float(input("Enter second number: "))

        if choice == '1':
            print(f"The result is: {add(num1, num2)}")
        elif choice == '2':
            print(f"The result is: {subtract(num1, num2)}")
        elif choice == '3':
            print(f"The result is: {multiply(num1, num2)}")
        elif choice == '4':
            print(f"The result is: {divide(num1, num2)}")
    else:
        print("Invalid Input")

if __name__ == "__main__":
    calculator()


Select operation:
1. Add
2. Subtract
3. Multiply
4. Divide
Enter choice(1/2/3/4): 2
Enter first number: 45
Enter second number: 5
The result is: 40.0


In [ ]:
#1.8. Write a program to find the larger of the three pre-specified numbers using ternary operators.

In [ ]:
# Define three numbers
num1 = 10
num2 = 20
num3 = 15

# Find the largest number using ternary operators
largest = (num1 if num1 > num2 else num2) if (num1 if num1 > num2 else num2) > num3 else num3

print(f"The largest number is: {largest}")


The largest number is: 20


In [ ]:
#1.9. Write a program to find the factors of a whole number using a while loop.

In [ ]:
def find_factors(number):
    i = 1
    factors = []

    while i <= number:
        if number % i == 0:
            factors.append(i)
        i += 1

    return factors

# Input: a whole number
num = int(input("Enter a whole number: "))

# Get the factors
factors = find_factors(num)

# Print the factors
print(f"The factors of {num} are: {factors}")


Enter a whole number: 40
The factors of 40 are: [1, 2, 4, 5, 8, 10, 20, 40]


In [ ]:
#1.10. Write a program to find the sum of all the positive numbers entered by the user.
#As soon as the user enters a negative number, stop taking in any further input from the user and display the sum.

In [ ]:
def sum_positive_numbers():
    total_sum = 0

    while True:
        num = float(input("Enter a number (negative number to stop): "))
        if num < 0:
            break
        total_sum += num

    print(f"The sum of all positive numbers entered is: {total_sum}")

# Run the function
sum_positive_numbers()


Enter a number (negative number to stop): 2
Enter a number (negative number to stop): 4
Enter a number (negative number to stop): 5
Enter a number (negative number to stop): 6
Enter a number (negative number to stop): -9
The sum of all positive numbers entered is: 17.0


In [ ]:
#1.11. Write a program to find prime numbers between 2 to 100 using nested for loops.

In [ ]:
def find_primes(start, end):
    primes = []
    for num in range(start, end + 1):
        if num > 1:  # All prime numbers are greater than 1
            is_prime = True
            for i in range(2, int(num ** 0.5) + 1):
                if num % i == 0:
                    is_prime = False
                    break
            if is_prime:
                primes.append(num)
    return primes

# Define the range
start = 2
end = 100

# Get the prime numbers
prime_numbers = find_primes(start, end)

# Print the prime numbers
print(f"Prime numbers between {start} and {end} are: {prime_numbers}")


Prime numbers between 2 and 100 are: [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97]


In [ ]:
#1.12. Write the programs for the following:
#Accept the marks of the student in five major subjects and display the same.
#Calculate the sum of the marks of all subjects.Divide the total marks by number of subjects (i.e. 5), calculate percentage = total marks/5 and display the percentage.
#Find the grade of the student as per the following criteria. Hint: Use Match & case for this.:
#Criteria                                                                    Grade
#percentage > 85                                                               A
#percentage < 85 && percentage >= 75                                           B
#percentage < 75 && percentage >= 50                                           C
#percentage > 30 && percentage <= 50                                           D
#percentage <30                                                                Reapper



In [ ]:
def calculate_grade(percentage):
    match percentage:
        case p if p > 85:
            return 'A'
        case p if 75 <= p < 85:
            return 'B'
        case p if 50 <= p < 75:
            return 'C'
        case p if 30 < p <= 50:
            return 'D'
        case p if p <= 30:
            return 'Reappear'

def main():
    # Accept marks for five major subjects
    marks = []
    for i in range(1, 6):
        mark = float(input(f"Enter the marks for subject {i}: "))
        marks.append(mark)

    # Display the marks
    print("\nMarks entered:")
    for i, mark in enumerate(marks, start=1):
        print(f"Subject {i}: {mark}")

    # Calculate total marks and percentage
    total_marks = sum(marks)
    percentage = total_marks / 5

    # Display total marks and percentage
    print(f"\nTotal Marks: {total_marks}")
    print(f"Percentage: {percentage:.2f}%")

    # Determine the grade
    grade = calculate_grade(percentage)
    print(f"Grade: {grade}")

if __name__ == "__main__":
    main()


Enter the marks for subject 1: 45
Enter the marks for subject 2: 67
Enter the marks for subject 3: 87
Enter the marks for subject 4: 80
Enter the marks for subject 5: 98

Marks entered:
Subject 1: 45.0
Subject 2: 67.0
Subject 3: 87.0
Subject 4: 80.0
Subject 5: 98.0

Total Marks: 377.0
Percentage: 75.40%
Grade: B


In [ ]:
#1.13. Write a program for VIBGYOR Spectrum based on their Wavelength using. Wavelength Range:
#COLOR					WAVELENGTH (nm)
#Violet						400.0-440.0
#Indigo					  440.0-460.0
#Blue						  460.0-500.0
#Green						500.0-570.0
#Yellow					  570.0-590.0
#Orange 					590.0-620.0
#Red						  620.0-720.0


In [ ]:
def determine_color(wavelength):
    match wavelength:
        case w if 400.0 <= w < 440.0:
            return 'Violet'
        case w if 440.0 <= w < 460.0:
            return 'Indigo'
        case w if 460.0 <= w < 500.0:
            return 'Blue'
        case w if 500.0 <= w < 570.0:
            return 'Green'
        case w if 570.0 <= w < 590.0:
            return 'Yellow'
        case w if 590.0 <= w < 620.0:
            return 'Orange'
        case w if 620.0 <= w <= 720.0:
            return 'Red'
        case _:
            return 'Wavelength out of range'

def main():
    # Accept wavelength input from user
    wavelength = float(input("Enter the wavelength (nm): "))

    # Determine the color
    color = determine_color(wavelength)

    # Display the color
    print(f"The color for wavelength {wavelength} nm is: {color}")

if __name__ == "__main__":
    main()


Enter the wavelength (nm): 560
The color for wavelength 560.0 nm is: Green


In [ ]:
#1.14.Consider the gravitational interactions between the Earth, Moon, and Sun in our solar system. Given:
#mass_earth = 5.972e24 # Mass of Earth in kilograms
#mass_moon = 7.34767309e22 # Mass of Moon in kilograms
#mass_sun = 1.989e30 # Mass of Sun in kilograms
#distance_earth_sun 1.496ell # Average distance between Earth and Sun in meters distance_moon_earth = 3.844e8 # Average distance between Moon and Earth in meters
#Tasks:
# Calculate the gravitational force between the Earth and the Sun.
#Calculate the gravitational force between the Moon and the Earth.
#Compare the calculated forces to determine which gravitational force is stronger.
#Explain which celestial body (Earth or Moon) is more attracted to the other based on the comparison.


In [ ]:
# Define the constants
G = 6.67430e-11  # Gravitational constant in m^3 kg^-1 s^-2
mass_earth = 5.972e24  # Mass of Earth in kg
mass_moon = 7.34767309e22  # Mass of Moon in kg
mass_sun = 1.989e30  # Mass of Sun in kg
distance_earth_sun = 1.496e11  # Average distance between Earth and Sun in m
distance_moon_earth = 3.844e8  # Average distance between Moon and Earth in m

# Calculate gravitational force between Earth and Sun
force_earth_sun = (G * mass_earth * mass_sun) / (distance_earth_sun ** 2)

# Calculate gravitational force between Moon and Earth
force_moon_earth = (G * mass_earth * mass_moon) / (distance_moon_earth ** 2)

# Compare the forces
stronger_force = "Earth and Sun" if force_earth_sun > force_moon_earth else "Moon and Earth"

# Print the results
print(f"Gravitational force between Earth and Sun: {force_earth_sun:.2e} N")
print(f"Gravitational force between Moon and Earth: {force_moon_earth:.2e} N")
print(f"The stronger gravitational force is between: {stronger_force}")

# Determine which celestial body is more attracted to the other
if force_earth_sun > force_moon_earth:
    print("The Earth is more attracted to the Sun.")
else:
    print("The Moon is more attracted to the Earth.")


Gravitational force between Earth and Sun: 3.54e+22 N
Gravitational force between Moon and Earth: 1.98e+20 N
The stronger gravitational force is between: Earth and Sun
The Earth is more attracted to the Sun.


In [ ]:
#2. Design and implement a Python program for managing student information using object-oriented principles. Create a class called "Student with encapsulated attributes for name, age, and roll number. Implement getter and setter methods for these attributes. Additionally, provide methods to display student information and update student details.
#Tasks:
#Define the 'Student class with encapsulated attributes.
#Implement getter and setter methods for the attributes.
#Write methods to display student information and update details.
#Create instances of the `Student class and test the implemented functionality.


In [ ]:
class Student:
    def __init__(self, name, age, roll_number):
        self.__name = name
        self.__age = age
        self.__roll_number = roll_number

    # Getter for name
    def get_name(self):
        return self.__name

    # Setter for name
    def set_name(self, name):
        self.__name = name

    # Getter for age
    def get_age(self):
        return self.__age

    # Setter for age
    def set_age(self, age):
        self.__age = age

    # Getter for roll number
    def get_roll_number(self):
        return self.__roll_number

    # Setter for roll number
    def set_roll_number(self, roll_number):
        self.__roll_number = roll_number

    # Method to display student information
    def display_info(self):
        print(f"Name: {self.__name}")
        print(f"Age: {self.__age}")
        print(f"Roll Number: {self.__roll_number}")

    # Method to update student details
    def update_details(self, name=None, age=None, roll_number=None):
        if name is not None:
            self.set_name(name)
        if age is not None:
            self.set_age(age)
        if roll_number is not None:
            self.set_roll_number(roll_number)

# Create instances of the Student class and test the functionality
student1 = Student("Alice", 20, "S12345")
student2 = Student("Bob", 22, "S67890")

# Display initial information
print("Initial Information:")
student1.display_info()
print()
student2.display_info()

# Update student1's details
student1.update_details(name="Alicia", age=21)
print("\nUpdated Information for Student 1:")
student1.display_info()

# Update student2's details
student2.update_details(roll_number="S98765")
print("\nUpdated Information for Student 2:")
student2.display_info()


Initial Information:
Name: Alice
Age: 20
Roll Number: S12345

Name: Bob
Age: 22
Roll Number: S67890

Updated Information for Student 1:
Name: Alicia
Age: 21
Roll Number: S12345

Updated Information for Student 2:
Name: Bob
Age: 22
Roll Number: S98765


In [ ]:
#3.Develop a Python program for managing library resources efficiently. Design a class named `LibraryBook` with attributes like book name, author, and availability status. Implement methods for borrowing and returning books while ensuring proper encapsulation of attributes.
#Tasks:
#1. Create the "LibraryBook" class with encapsulated attributes.
#2. Implement methods for borrowing and returning books.
#3. Ensure proper encapsulation to protect book details.
#4. Test the borrowing and returning functionality with sample data.


In [ ]:
class LibraryBook:
    def __init__(self, book_name, author):
        self.__book_name = book_name
        self.__author = author
        self.__is_available = True

    # Getter for book name
    def get_book_name(self):
        return self.__book_name

    # Getter for author
    def get_author(self):
        return self.__author

    # Getter for availability status
    def is_available(self):
        return self.__is_available

    # Method to borrow the book
    def borrow_book(self):
        if self.__is_available:
            self.__is_available = False
            print(f"You have successfully borrowed '{self.__book_name}'.")
        else:
            print(f"Sorry, '{self.__book_name}' is currently not available.")

    # Method to return the book
    def return_book(self):
        if not self.__is_available:
            self.__is_available = True
            print(f"You have successfully returned '{self.__book_name}'.")
        else:
            print(f"'{self.__book_name}' is already available in the library.")

    # Method to display book details
    def display_info(self):
        availability = "Available" if self.__is_available else "Not Available"
        print(f"Book Name: {self.__book_name}")
        print(f"Author: {self.__author}")
        print(f"Availability: {availability}")

# Create instances of the LibraryBook class and test the functionality
book1 = LibraryBook("1984", "George Orwell")
book2 = LibraryBook("To Kill a Mockingbird", "Harper Lee")

# Display initial information
print("Initial Information:")
book1.display_info()
print()
book2.display_info()

# Borrow book1
print("\nBorrowing '1984':")
book1.borrow_book()
print("\nTrying to borrow '1984' again:")
book1.borrow_book()

# Return book1
print("\nReturning '1984':")
book1.return_book()
print("\nTrying to return '1984' again:")
book1.return_book()

# Display updated information
print("\nUpdated Information:")
book1.display_info()
print()
book2.display_info()


Initial Information:
Book Name: 1984
Author: George Orwell
Availability: Available

Book Name: To Kill a Mockingbird
Author: Harper Lee
Availability: Available

Borrowing '1984':
You have successfully borrowed '1984'.

Trying to borrow '1984' again:
Sorry, '1984' is currently not available.

Returning '1984':
You have successfully returned '1984'.

Trying to return '1984' again:
'1984' is already available in the library.

Updated Information:
Book Name: 1984
Author: George Orwell
Availability: Available

Book Name: To Kill a Mockingbird
Author: Harper Lee
Availability: Available


In [ ]:
#4.Create a simple banking system using object-oriented concepts in Python. Design classes representing different types of bank accounts such as savings and checking. Implement methods for deposit, withdraw, and balance inquiry. Utilize inheritance to manage different account types efficiently.
#Tasks:
#1. Define base class(es) for bank accounts with common attributes and methods.
#2. Implement subclasses for specific account types (e.g., Savings Account, Checking Account).
#3. Provide methods for deposit, withdraw, and balance inquiry in each subclass.
#4. Test the banking system by creating instances of different account types and performing transactions.


In [ ]:
class BankAccount:
    def __init__(self, account_number, balance=0):
        self.account_number = account_number
        self.balance = balance

    def deposit(self, amount):
        if amount > 0:
            self.balance += amount
            print(f"Deposited {amount}. New balance: {self.balance}")
        else:
            print("Deposit amount must be positive.")

    def withdraw(self, amount):
        if 0 < amount <= self.balance:
            self.balance -= amount
            print(f"Withdrew {amount}. New balance: {self.balance}")
        else:
            print("Insufficient funds or invalid withdrawal amount.")

    def get_balance(self):
        print(f"Account balance: {self.balance}")
        return self.balance


class SavingsAccount(BankAccount):
    def __init__(self, account_number, balance=0, interest_rate=0.02):
        super().__init__(account_number, balance)
        self.interest_rate = interest_rate

    def add_interest(self):
        interest = self.balance * self.interest_rate
        self.deposit(interest)
        print(f"Interest added at rate {self.interest_rate}. New balance: {self.balance}")


class CheckingAccount(BankAccount):
    def __init__(self, account_number, balance=0, overdraft_limit=500):
        super().__init__(account_number, balance)
        self.overdraft_limit = overdraft_limit

    def withdraw(self, amount):
        if 0 < amount <= self.balance + self.overdraft_limit:
            self.balance -= amount
            print(f"Withdrew {amount}. New balance: {self.balance}")
        else:
            print("Overdraft limit exceeded or invalid withdrawal amount.")


# Test the banking system
savings = SavingsAccount("S12345", 1000)
checking = CheckingAccount("C67890", 500)

# Display initial balances
print("Initial Balances:")
savings.get_balance()
checking.get_balance()

# Perform transactions on SavingsAccount
print("\nTransactions for Savings Account:")
savings.deposit(500)
savings.withdraw(200)
savings.add_interest()
savings.get_balance()

# Perform transactions on CheckingAccount
print("\nTransactions for Checking Account:")
checking.deposit(300)
checking.withdraw(1000)
checking.get_balance()


Initial Balances:
Account balance: 1000
Account balance: 500

Transactions for Savings Account:
Deposited 500. New balance: 1500
Withdrew 200. New balance: 1300
Deposited 26.0. New balance: 1326.0
Interest added at rate 0.02. New balance: 1326.0
Account balance: 1326.0

Transactions for Checking Account:
Deposited 300. New balance: 800
Withdrew 1000. New balance: -200
Account balance: -200


-200

In [ ]:
#5.Write a Python program that models different animals and their sounds. Design a base class called "Animal" with a method `make_sound()*. Create subclasses like "Dog" and "Cat" that override the `make_sound() method to produce appropriate sounds.
#Tasks:
#1. Define the "Animal class with a method `make_sound().
#2. Create subclasses "Dog" and "Cat" that override the "make_sound() method.
#3. Implement the sound generation logic for each subclass.
#4. Test the program by creating instances of "Dog" and "Cat" and calling the `make_sound()` method.


In [ ]:
class Animal:
    def make_sound(self):
        raise NotImplementedError("Subclass must implement abstract method")

class Dog(Animal):
    def make_sound(self):
        return "Woof!"

class Cat(Animal):
    def make_sound(self):
        return "Meow!"

# Test the program by creating instances of Dog and Cat and calling the make_sound() method
def test_animal_sounds():
    animals = [Dog(), Cat()]
    for animal in animals:
        print(f"The {animal.__class__.__name__} says: {animal.make_sound()}")

test_animal_sounds()


The Dog says: Woof!
The Cat says: Meow!


In [ ]:
#6.Write a code for Restaurant Management System Using OOPS:
#Create a Menultem class that has attributes such as name, description, price, and category.
#Implement methods to add a new menu item, update menu item information, and remove a menu item from the menu.
#Use encapsulation to hide the menu item's unique identification number.
#Inherit from the Menultem class to create a FoodItem class and a Beverageltem class, each with their own specific attributes and methods.

In [ ]:
# Base class for Menu Items
class MenuItem:
    def __init__(self, name, description, price, category):
        self.__id = id(self)  # Encapsulated unique ID
        self.name = name
        self.description = description
        self.price = price
        self.category = category

    def update_item(self, name=None, description=None, price=None, category=None):
        if name: self.name = name
        if description: self.description = description
        if price: self.price = price
        if category: self.category = category

    def __str__(self):
        return f"{self.name}: {self.description}, Price: {self.price}, Category: {self.category}"

# Derived class for Food Items
class FoodItem(MenuItem):
    def __init__(self, name, description, price, category, cuisine_type):
        super().__init__(name, description, price, category)
        self.cuisine_type = cuisine_type

# Derived class for Beverage Items
class BeverageItem(MenuItem):
    def __init__(self, name, description, price, category, size):
        super().__init__(name, description, price, category)
        self.size = size

# Menu Management
class Menu:
    def __init__(self):
        self.items = []

    def add_item(self, item):
        self.items.append(item)

    def remove_item(self, item):
        self.items.remove(item)

    def display_menu(self):
        for item in self.items:
            print(item)

# Example Usage
menu = Menu()
burger = FoodItem("Burger", "Beef burger", 5.99, "Main Course", "American")
coffee = BeverageItem("Coffee", "Hot coffee", 2.99, "Beverage", "Medium")

menu.add_item(burger)
menu.add_item(coffee)
menu.display_menu()

# Update and Remove Example
burger.update_item(price=6.99)
menu.remove_item(coffee)
menu.display_menu()


Burger: Beef burger, Price: 5.99, Category: Main Course
Coffee: Hot coffee, Price: 2.99, Category: Beverage
Burger: Beef burger, Price: 6.99, Category: Main Course


In [ ]:
#7.Write a code for Hotel Management System using OOPS:
#Create a Room class that has attributes such as room number, room type, rate, and availability (private). Implement methods to book a room, check in a guest, and check out a guest.
#Use encapsulation to hide the room's unique identification number.
#Inherit from the Room class to create a SuiteRoom class and a Standard Room class, each with their own specific attributes and methods.

In [ ]:
# Base class for Rooms
class Room:
    def __init__(self, room_number, room_type, rate):
        self.__id = id(self)  # Encapsulated unique ID
        self.room_number = room_number
        self.room_type = room_type
        self.rate = rate
        self.__availability = True  # Private attribute

    def book_room(self):
        if self.__availability:
            self.__availability = False
            return True
        return False

    def check_in(self):
        if not self.__availability:
            print(f"Room {self.room_number} is already booked.")
        else:
            self.book_room()
            print(f"Guest checked into Room {self.room_number}.")

    def check_out(self):
        if self.__availability:
            print(f"Room {self.room_number} is already vacant.")
        else:
            self.__availability = True
            print(f"Guest checked out from Room {self.room_number}.")

    def __str__(self):
        status = "Available" if self.__availability else "Booked"
        return f"Room {self.room_number}, Type: {self.room_type}, Rate: {self.rate}, Status: {status}"

# Derived class for Suite Rooms
class SuiteRoom(Room):
    def __init__(self, room_number, rate, amenities):
        super().__init__(room_number, "Suite", rate)
        self.amenities = amenities

# Derived class for Standard Rooms
class StandardRoom(Room):
    def __init__(self, room_number, rate):
        super().__init__(room_number, "Standard", rate)

# Hotel Management
class Hotel:
    def __init__(self):
        self.rooms = []

    def add_room(self, room):
        self.rooms.append(room)

    def find_room(self, room_number):
        for room in self.rooms:
            if room.room_number == room_number:
                return room
        return None

    def display_rooms(self):
        for room in self.rooms:
            print(room)

# Example Usage
hotel = Hotel()
suite = SuiteRoom(101, 250.00, ["Jacuzzi", "Mini Bar"])
standard = StandardRoom(102, 100.00)

hotel.add_room(suite)
hotel.add_room(standard)
hotel.display_rooms()

# Check-in and Check-out Example
room = hotel.find_room(101)
room.check_in()
room.check_out()


Room 101, Type: Suite, Rate: 250.0, Status: Available
Room 102, Type: Standard, Rate: 100.0, Status: Available
Guest checked into Room 101.
Guest checked out from Room 101.


In [ ]:
#8.Write a code for Fitness Club Management System using OOPS:
#Create a Member class that has attributes such as name, age, membership type, and membership status (private).
#Implement methods to register a new member, renew a membership, and cancel a membership.
#Use encapsulation to hide the member's unique identification number.
#Inherit from the Member class to create a Family Member class and an IndividualMember class, each with their own specific attributes and methods.

In [ ]:
# Base class for Club Members
class Member:
    def __init__(self, name, age, membership_type):
        self.__id = id(self)  # Encapsulated unique ID
        self.name = name
        self.age = age
        self.membership_type = membership_type
        self.__membership_status = "Active"  # Private attribute

    def renew_membership(self):
        if self.__membership_status == "Active":
            print(f"Membership for {self.name} is already active.")
        else:
            self.__membership_status = "Active"
            print(f"Membership for {self.name} has been renewed.")

    def cancel_membership(self):
        if self.__membership_status == "Cancelled":
            print(f"Membership for {self.name} is already cancelled.")
        else:
            self.__membership_status = "Cancelled"
            print(f"Membership for {self.name} has been cancelled.")

    def __str__(self):
        return f"Name: {self.name}, Age: {self.age}, Membership Type: {self.membership_type}, Status: {self.__membership_status}"

# Derived class for Family Members
class FamilyMember(Member):
    def __init__(self, name, age, membership_type, family_members):
        super().__init__(name, age, membership_type)
        self.family_members = family_members

    def add_family_member(self, member_name):
        self.family_members.append(member_name)

# Derived class for Individual Members
class IndividualMember(Member):
    def __init__(self, name, age, membership_type, personal_trainer):
        super().__init__(name, age, membership_type)
        self.personal_trainer = personal_trainer

# Fitness Club Management
class FitnessClub:
    def __init__(self):
        self.members = []

    def register_member(self, member):
        self.members.append(member)
        print(f"{member.name} has been registered.")

    def display_members(self):
        for member in self.members:
            print(member)

# Example Usage
club = FitnessClub()
john = IndividualMember("John Doe", 30, "Individual", "Trainer Mike")
smith_family = FamilyMember("Smith Family", 40, "Family", ["Jane", "Bob"])

club.register_member(john)
club.register_member(smith_family)
club.display_members()

# Renew and Cancel Example
john.cancel_membership()
smith_family.renew_membership()
club.display_members()


John Doe has been registered.
Smith Family has been registered.
Name: John Doe, Age: 30, Membership Type: Individual, Status: Active
Name: Smith Family, Age: 40, Membership Type: Family, Status: Active
Membership for John Doe has been cancelled.
Membership for Smith Family is already active.
Name: John Doe, Age: 30, Membership Type: Individual, Status: Cancelled
Name: Smith Family, Age: 40, Membership Type: Family, Status: Active


In [ ]:
#9. Write a code for Event Management System using OOPS:
#Create an Event class that has attributes such as name, date, time, location, and list of attendees (private). Implement methods to create a new event, add or remove attendees, and get the total number of attendees.
#Use encapsulation to hide the event's unique identification number.
#Inherit from the Event class to create a PrivateEvent class and a PublicEvent class, each with their own specific attributes and methods.

In [ ]:
# Base class for Events
class Event:
    def __init__(self, name, date, time, location):
        self.__id = id(self)  # Encapsulated unique ID
        self.name = name
        self.date = date
        self.time = time
        self.location = location
        self.__attendees = []  # Private list of attendees

    def add_attendee(self, attendee_name):
        self.__attendees.append(attendee_name)

    def remove_attendee(self, attendee_name):
        if attendee_name in self.__attendees:
            self.__attendees.remove(attendee_name)
        else:
            print(f"{attendee_name} is not in the attendee list.")

    def get_total_attendees(self):
        return len(self.__attendees)

    def __str__(self):
        return f"Event: {self.name}, Date: {self.date}, Time: {self.time}, Location: {self.location}, Attendees: {self.get_total_attendees()}"

# Derived class for Private Events
class PrivateEvent(Event):
    def __init__(self, name, date, time, location, host_name):
        super().__init__(name, date, time, location)
        self.host_name = host_name

    def send_invitation(self):
        print(f"Invitation sent by {self.host_name} for the event {self.name}.")

# Derived class for Public Events
class PublicEvent(Event):
    def __init__(self, name, date, time, location, entry_fee):
        super().__init__(name, date, time, location)
        self.entry_fee = entry_fee

    def promote_event(self):
        print(f"Public event {self.name} is being promoted. Entry fee: ${self.entry_fee}")

# Event Management
class EventManager:
    def __init__(self):
        self.events = []

    def create_event(self, event):
        self.events.append(event)
        print(f"Event '{event.name}' has been created.")

    def display_events(self):
        for event in self.events:
            print(event)

# Example Usage
manager = EventManager()
wedding = PrivateEvent("Wedding", "2024-09-15", "5:00 PM", "Beach Resort", "Alice")
concert = PublicEvent("Concert", "2024-08-30", "7:00 PM", "City Park", 20.00)

manager.create_event(wedding)
manager.create_event(concert)

wedding.add_attendee("John")
wedding.add_attendee("Mary")
concert.add_attendee("Chris")

manager.display_events()

# Example of Event-specific Methods
wedding.send_invitation()
concert.promote_event()


Event 'Wedding' has been created.
Event 'Concert' has been created.
Event: Wedding, Date: 2024-09-15, Time: 5:00 PM, Location: Beach Resort, Attendees: 2
Event: Concert, Date: 2024-08-30, Time: 7:00 PM, Location: City Park, Attendees: 1
Invitation sent by Alice for the event Wedding.
Public event Concert is being promoted. Entry fee: $20.0


In [ ]:
#10.Write a code for Airline Reservation System using OOPS:
#Create a Flight class that has attributes such as flight number, departure and arrival airports, departure and arrival times, and available seats (private).
#Implement methods to book a seat, cancel a reservation, and get the remaining available seats.
#Use encapsulation to hide the flight's unique identification number.
#Inherit from the Flight class to create a Domestic Flight class and an InternationalFlight class, each with their own specific attributes and methods.

In [ ]:
# Base class for Flights
class Flight:
    def __init__(self, flight_number, departure_airport, arrival_airport, departure_time, arrival_time, total_seats):
        self.__id = id(self)  # Encapsulated unique ID
        self.flight_number = flight_number
        self.departure_airport = departure_airport
        self.arrival_airport = arrival_airport
        self.departure_time = departure_time
        self.arrival_time = arrival_time
        self.__available_seats = total_seats  # Private attribute

    def book_seat(self):
        if self.__available_seats > 0:
            self.__available_seats -= 1
            print(f"Seat booked on flight {self.flight_number}.")
        else:
            print("No available seats.")

    def cancel_reservation(self):
        self.__available_seats += 1
        print(f"Reservation cancelled on flight {self.flight_number}.")

    def get_remaining_seats(self):
        return self.__available_seats

    def __str__(self):
        return f"Flight {self.flight_number} from {self.departure_airport} to {self.arrival_airport}, Departure: {self.departure_time}, Arrival: {self.arrival_time}, Seats Available: {self.get_remaining_seats()}"

# Derived class for Domestic Flights
class DomesticFlight(Flight):
    def __init__(self, flight_number, departure_airport, arrival_airport, departure_time, arrival_time, total_seats, state_tax):
        super().__init__(flight_number, departure_airport, arrival_airport, departure_time, arrival_time, total_seats)
        self.state_tax = state_tax

# Derived class for International Flights
class InternationalFlight(Flight):
    def __init__(self, flight_number, departure_airport, arrival_airport, departure_time, arrival_time, total_seats, passport_required):
        super().__init__(flight_number, departure_airport, arrival_airport, departure_time, arrival_time, total_seats)
        self.passport_required = passport_required

# Airline Reservation System Management
class AirlineReservationSystem:
    def __init__(self):
        self.flights = []

    def add_flight(self, flight):
        self.flights.append(flight)
        print(f"Flight {flight.flight_number} added to the system.")

    def display_flights(self):
        for flight in self.flights:
            print(flight)

# Example Usage
system = AirlineReservationSystem()
flight1 = DomesticFlight("AI101", "JFK", "LAX", "08:00 AM", "11:00 AM", 150, state_tax=8.25)
flight2 = InternationalFlight("AI202", "JFK", "LHR", "06:00 PM", "06:00 AM", 200, passport_required=True)

system.add_flight(flight1)
system.add_flight(flight2)

flight1.book_seat()
flight2.book_seat()
flight2.cancel_reservation()

system.display_flights()


Flight AI101 added to the system.
Flight AI202 added to the system.
Seat booked on flight AI101.
Seat booked on flight AI202.
Reservation cancelled on flight AI202.
Flight AI101 from JFK to LAX, Departure: 08:00 AM, Arrival: 11:00 AM, Seats Available: 149
Flight AI202 from JFK to LHR, Departure: 06:00 PM, Arrival: 06:00 AM, Seats Available: 200


In [ ]:
#11. Define a Python module named constants.py containing constants like pi and the speed of light.

In [ ]:
# constants.py

# Mathematical constants
PI = 3.141592653589793

# Physical constants
SPEED_OF_LIGHT = 299792458  # in meters per second
GRAVITATIONAL_CONSTANT = 6.67430e-11  # in m^3 kg^-1 s^-2

# Other useful constants
AVOGADRO_NUMBER = 6.02214076e23  # in mol^-1
PLANCK_CONSTANT = 6.62607015e-34  # in m^2 kg / s


In [ ]:
#12. Write a Python module named calculator.py containing functions for addition, subtraction, multiplication, and division.

In [ ]:
# calculator.py

def add(a, b):
    """Returns the sum of a and b."""
    return a + b

def subtract(a, b):
    """Returns the difference of a and b."""
    return a - b

def multiply(a, b):
    """Returns the product of a and b."""
    return a * b

def divide(a, b):
    """Returns the quotient of a and b. Raises an error if b is zero."""
    if b == 0:
        raise ValueError("Cannot divide by zero.")
    return a / b


In [ ]:
#13. Implement a Python package structure for a project named ecommerce, containing modules for product management and order processing.

In [ ]:
# constants.py

# Mathematical constants
PI = 3.141592653589793

# Physical constants
SPEED_OF_LIGHT = 299792458  # in meters per second
GRAVITATIONAL_CONSTANT = 6.67430e-11  # in m^3 kg^-1 s^-2

# Other useful constants
AVOGADRO_NUMBER = 6.02214076e23  # in mol^-1
PLANCK_CONSTANT = 6.62607015e-34  # in m^2 kg / s


In [ ]:
#14. Implement a Python module named string_utils.py containing functions for string manipulation, such as reversing and capitalizing strings.

In [ ]:
# string_utils.py

def reverse_string(s):
    """Returns the reversed version of the input string."""
    return s[::-1]

def capitalize_string(s):
    """Returns the string with the first letter capitalized."""
    return s.capitalize()

def to_uppercase(s):
    """Returns the string in uppercase."""
    return s.upper()

def to_lowercase(s):
    """Returns the string in lowercase."""
    return s.lower()


In [ ]:
#15. Write a Python module named file_operations.py with functions for reading, writing, and appending data to a file.

In [ ]:
# file_operations.py

def write_to_file(file_path, data):
    """Writes the provided data to the specified file, overwriting any existing content."""
    with open(file_path, 'w') as file:
        file.write(data)

def append_to_file(file_path, data):
    """Appends the provided data to the specified file."""
    with open(file_path, 'a') as file:
        file.write(data)

def read_from_file(file_path):
    """Reads and returns the content of the specified file."""
    with open(file_path, 'r') as file:
        return file.read()


In [ ]:
#16. Write a Python program to create a text file named "employees.txt" and write the details of employees, including their name, age, and salary, into the file.

In [ ]:
# Define employee details
employees = [
    {"name": "Alice", "age": 30, "salary": 70000},
    {"name": "Bob", "age": 25, "salary": 50000},
    {"name": "Charlie", "age": 35, "salary": 80000}
]

# File path
file_path = "employees.txt"

# Write employee details to the file
with open(file_path, 'w') as file:
    for employee in employees:
        line = f"Name: {employee['name']}, Age: {employee['age']}, Salary: ${employee['salary']}\n"
        file.write(line)

print(f"Employee details have been written to {file_path}.")


Employee details have been written to employees.txt.


In [ ]:
#17. Develop a Python script that opens an existing text file named "inventory.txt" in read mode and displays the contents of the file line by line.

In [ ]:
# File path
file_path = "inventory.txt"

# Open the file and read line by line
try:
    with open(file_path, 'r') as file:
        for line in file:
            print(line, end='')  # Print each line without adding extra newlines
except FileNotFoundError:
    print(f"The file {file_path} does not exist.")


The file inventory.txt does not exist.


In [ ]:
#18. Create a Python script that reads a text file named "expenses.txt" and calculates the total amount spent on various expenses listed in the file.

In [ ]:
# File path
file_path = "expenses.txt"

# Initialize total amount
total_amount = 0.0

# Open the file and process each line
try:
    with open(file_path, 'r') as file:
        for line in file:
            try:
                # Convert line to float and add to total_amount
                amount = float(line.strip())
                total_amount += amount
            except ValueError:
                print(f"Skipping invalid line: {line.strip()}")

    print(f"Total amount spent: ${total_amount:.2f}")

except FileNotFoundError:
    print(f"The file {file_path} does not exist.")


The file expenses.txt does not exist.


In [ ]:
#19. Create a Python program that reads a text file named "paragraph.txt" and counts the occurrences of each word in the paragraph, displaying the results in alphabetical order.

In [ ]:
from collections import Counter
import re

# File path
file_path = "paragraph.txt"

# Initialize a Counter to count word occurrences
word_count = Counter()

# Read the file and process its content
try:
    with open(file_path, 'r') as file:
        text = file.read()
        # Use regex to find words and convert them to lowercase
        words = re.findall(r'\b\w+\b', text.lower())
        word_count.update(words)

    # Display word counts in alphabetical order
    for word in sorted(word_count):
        print(f"{word}: {word_count[word]}")

except FileNotFoundError:
    print(f"The file {file_path} does not exist.")


The file paragraph.txt does not exist.


In [ ]:
#20. What do you mean by Measure of Central Tendency and Measures of Dispersion .How it can be calculated.

Measures of Central Tendency
Mean: Average value.
Mean
=
∑
𝑥
𝑖
𝑛
Mean=
n
∑x
i
​

​

Median: Middle value when sorted.
Mode: Most frequent value.
Measures of Dispersion
Range: Difference between max and min.
Range
=
Max
−
Min
Range=Max−Min
Variance: Average of squared differences from the mean.
Variance
=
∑
(
𝑥
𝑖
−
Mean
)
2
𝑛
Variance=
n
∑(x
i
​
 −Mean)
2

​

Standard Deviation: Square root of variance.
IQR: Range between the 1st and 3rd quartiles.
IQR
=
𝑄
3
−
𝑄
1
IQR=Q3−Q1

In [ ]:
#21. What do you mean by skewness.Explain its types.Use graph to show.

Skewness
Skewness measures the asymmetry of a data distribution.

Types
Positive Skew (Right Skew):

Tail: Right side longer.
Graph: Peak on left, tail on right.
Negative Skew (Left Skew):

Tail: Left side longer.
Graph: Peak on right, tail on left.
Symmetric:

Tail: Even on both sides.
Graph: Bell-shaped or uniform.
Graphs:

Positive Skew: Tail on the right.
Negative Skew: Tail on the left.
Symmetric: Even distribution.

In [ ]:
#22. Explain PROBABILITY MASS FUNCTION (PMF) and PROBABILITY DENSITY FUNCTION (PDF). and what is the difference between them?

PMF (Probability Mass Function)
For: Discrete random variables.
Gives: Probability of exact outcomes.
Sum: Total probabilities = 1.
PDF (Probability Density Function)
For: Continuous random variables.
Gives: Density of outcomes; probabilities are found by integrating.
Area: Total area under the curve = 1.
Difference: PMF provides exact probabilities for discrete values; PDF provides density, and probabilities are found over intervals by integration.

In [ ]:
#23. What is correlation. Explain its type in details.what are the methods of determining correlation

Correlation
Definition: Measures the linear relationship between two variables.
Types
Positive Correlation: Both variables increase together.
𝑟
r > 0.
Negative Correlation: One variable increases as the other decreases.
𝑟
r < 0.
No Correlation: No linear relationship.
𝑟
r ≈ 0.
Methods
Pearson: For linear relationships.
𝑟
r ranges from -1 to +1.
Spearman's Rank: For non-linear but monotonic relationships or ordinal data.
𝜌
ρ ranges from -1 to +1.
Kendall's Tau: For ordinal data and monotonic relationships.
𝜏
τ ranges from -1 to +1.

In [ ]:
#24. Calculate coefficient of correlation between the marks obtained by 10 students in Accountancy and statistics:
#Student     1     2     3     4     5    6     7     8     9     10
#Accountancy 45    70    65    30    90   40   50     75    85   60
#Statistics  35    90    70    40    95   40   60    80     80   50
#Use Karl Pearson's Coefficient of Correlation Method to find it.

To calculate the Pearson correlation coefficient
𝑟
r:

Data
Accountancy (X): 45, 70, 65, 30, 90, 40, 50, 75, 85, 60
Statistics (Y): 35, 90, 70, 40, 95, 40, 60, 80, 80, 50
Formula
𝑟
=
𝑛
∑
𝑋
𝑌
−
(
∑
𝑋
)
(
∑
𝑌
)
[
𝑛
∑
𝑋
2
−
(
∑
𝑋
)
2
]
[
𝑛
∑
𝑌
2
−
(
∑
𝑌
)
2
]
r=
[n∑X
2
 −(∑X)
2
 ][n∑Y
2
 −(∑Y)
2
 ]
​

n∑XY−(∑X)(∑Y)
​

Steps
Calculate Sums:

∑
𝑋
=
625
∑X=625
∑
𝑌
=
660
∑Y=660
∑
𝑋
𝑌
=
42
,
050
∑XY=42,050
∑
𝑋
2
=
39
,
250
∑X
2
 =39,250
∑
𝑌
2
=
43
,
350
∑Y
2
 =43,350
Apply Formula:

𝑟
=
10
×
42
,
050
−
(
625
×
660
)
[
10
×
39
,
250
−
62
5
2
]
[
10
×
43
,
350
−
66
0
2
]
r=
[10×39,250−625
2
 ][10×43,350−660
2
 ]
​

10×42,050−(625×660)
​


𝑟
≈
0.94
r≈0.94
Result
The Pearson correlation coefficient
𝑟
r is approximately 0.94, indicating a strong positive relationship.

In [ ]:
#25. Discuss the 4 differences between correlation and regression.

Here’s a more concise comparison:

Purpose:

Correlation measures the strength and direction of a relationship between two variables.
Regression predicts the value of one variable based on another.
Directionality:

Correlation is the same in both directions.
Regression distinguishes between dependent and independent variables.
Measurement:

Correlation provides a single coefficient between -1 and 1.
Regression gives an equation for predicting values.
Scope:

Correlation shows relationship strength.
Regression allows prediction and assesses impact.

In [ ]:
#26. Find the most likely price at Delhi corresponding to the price of Rs. 70 at Agra from the following data: Coefficient of correlation between the prices of the two places +0.8.

To find the most likely price at Delhi corresponding to a price of Rs. 70 at Agra, you'll need the following data:

Mean price at Agra (μ₁) and Mean price at Delhi (μ₂).
Standard deviation of prices at Agra (σ₁) and Standard deviation of prices at Delhi (σ₂).
The formula to calculate the most likely price at Delhi (
𝑋
2
X
2
​
 ) is:

𝑋
2
=
𝜇
2
+
𝑟
×
𝜎
2
𝜎
1
×
(
𝑋
1
−
𝜇
1
)
X
2
​
 =μ
2
​
 +r×
σ
1
​

σ
2
​

​
 ×(X
1
​
 −μ
1
​
 )
Where:

𝑟
r is the correlation coefficient (+0.8).
𝑋
1
X
1
​
  is the given price at Agra (Rs. 70).
However, since the mean prices and standard deviations are not provided, we can't calculate the exact price at Delhi. If these values are available, you can use the formula above to find the answer.

In [ ]:
#27. In a partially destroyed laboratory record of an analysis of correlation data, the following results only are legible: Variance of x = 9, Regression equations are: (i) 8x-10y=-66; (ii) 40x - 18y = 214. What are (a) the mean values of x and y, (b) the coefficient of correlation between x and y, (c) the σ of y.

Given:
Variance of
𝑥
=
9
x=9 (i.e.,
𝜎
𝑥
=
3
σ
x
​
 =3)
Regression equations:
8
𝑥
−
10
𝑦
=
−
66
8x−10y=−66
40
𝑥
−
18
𝑦
=
214
40x−18y=214
(a) Mean Values of
𝑥
x and
𝑦
y:
At the means, both regression equations intersect:

4
5
𝑥
‾
+
33
5
=
20
9
𝑥
‾
−
107
9
5
4
​
  
x
 +
5
33
​
 =
9
20
​
  
x
 −
9
107
​

Solve for
𝑥
‾
x
 , then substitute to find
𝑦
‾
y
​
 .

(b) Coefficient of Correlation (
𝑟
r):
𝑟
=
4
5
×
20
9
=
80
45
=
16
9
=
4
3
r=
5
4
​
 ×
9
20
​

​
 =
45
80
​

​
 =
9
16
​

​
 =
3
4
​

(c) Standard Deviation of
𝑦
y (
𝜎
𝑦
σ
y
​
 ):
𝜎
𝑦
=
3
×
20
9
=
60
9
=
20
3
σ
y
​
 =3×
9
20
​
 =
9
60
​
 =
3
20
​


In [ ]:
#28. What is Normal Distribution? What are the four Assumptions of Normal Distribution? Explain in detail.

Normal Distribution:
A Normal Distribution is a bell-shaped, symmetrical curve where the mean, median, and mode are equal. It's used to describe data that clusters around a central value.

Four Assumptions:
Linearity: The relationship between variables is linear.
Independence: Observations are independent of each other.
Homoscedasticity: Variance is constant across all levels of variables.
Normality: Data errors follow a normal distribution.







In [ ]:
#29. Write all the characteristics or Properties of the Normal Distribution Curve.

The characteristics of the Normal Distribution Curve are:

Bell-shaped Curve: Symmetrical around the mean.
Mean, Median, Mode: All are equal and located at the center.
Asymptotic: The curve never touches the x-axis.
Unimodal: Has a single peak.
68-95-99.7 Rule: About 68% of data lies within 1 standard deviation, 95% within 2, and 99.7% within 3 standard deviations from the mean.
Total Area: The area under the curve equals 1.
Symmetry: Divided into two equal halves by the mean.
Continuous: Values range from negative to positive infinity.








In [ ]:
#30. Which of the following options are correct about Normal Distribution Curve.
#(a) Within a range 0.6745 of o on both sides the middle 50% of the observations occur i,e, mean ±0.67450 covers 50% area 25% on each side.
#(b) Mean ±1S.D. (i,e.ulo) covers 68.268% area, 34.134 % area lies on either side of the mean.
#(c) Mean ±2S.D. (i,e. μ +20) covers 95.45% area, 47.725% area lies on either side of the mean.
#(d) Mean ±3 S.D. (i,e. μ ±30) covers 99.73% area, 49.856% area lies on the either side of the mean. (e) Only 0.27% area is outside the range μ ±30.

The correct options about the Normal Distribution Curve are:

(b) Mean ±1S.D. (i.e., μ ± σ) covers 68.268% area, 34.134% area lies on either side of the mean.
(c) Mean ±2S.D. (i.e., μ ± 2σ) covers 95.45% area, 47.725% area lies on either side of the mean.
(d) Mean ±3 S.D. (i.e., μ ± 3σ) covers 99.73% area, 49.865% area lies on either side of the mean.
(e) Only 0.27% area is outside the range μ ± 3σ.

In [ ]:
#31. The mean of a distribution is 60 with a standard deviation of 10. Assuming that the distribution is normal, what percentage of items be (i) between 60 and 72, (ii) between 50 and 60, (iii) beyond 72 and (iv) between
70 and 80?

Given a normal distribution with a mean (μ) of 60 and a standard deviation (σ) of 10:

Between 60 and 72:
𝑧
=
72
−
60
10
=
1.2
z=
10
72−60
​
 =1.2

Percentage: 38.21%
Between 50 and 60:
𝑧
=
60
−
50
10
=
1
z=
10
60−50
​
 =1

Percentage: 34.13%
Beyond 72:

Percentage: 11.51%
Between 70 and 80:

Percentage: 18.41%

In [ ]:
#32. 15000 students sat for an examination. The mean marks was 49 and the distribution of marks had a standard deviation of 6. Assuming that the marks were normally distributed what proportion of students scored (a) more than 55 marks, (b) more than 70 marks

Given a normal distribution with a mean (μ) of 49 and a standard deviation (σ) of 6:

Proportion of students who scored more than 55 marks:

𝑧
=
55
−
49
6
=
1
z=
6
55−49
​
 =1
Proportion: 15.87%
Proportion of students who scored more than 70 marks:

𝑧
=
70
−
49
6
=
3.5
z=
6
70−49
​
 =3.5
Proportion: 0.02%

In [ ]:
#33. If the height of 500 students are normally distributed with mean 65 inch and standard deviation 5 inch. How many students have height: a) greater than 70 inch. b) between 60 and 70 inch.

Given a normal distribution with a mean (μ) of 65 inches and a standard deviation (σ) of 5 inches for 500 students:

Students with height greater than 70 inches:

𝑧
=
70
−
65
5
=
1
z=
5
70−65
​
 =1
Proportion above 70 inches: 15.87%
Number of students:
500
×
0.1587
=
79
500×0.1587=79 students.
Students with height between 60 and 70 inches:

For 60 inches:
𝑧
=
60
−
65
5
=
−
1
z=
5
60−65
​
 =−1
For 70 inches:
𝑧
=
70
−
65
5
=
1
z=
5
70−65
​
 =1
Proportion between 60 and 70 inches: 68.27%
Number of students:
500
×
0.6827
=
341
500×0.6827=341 students.

In [ ]:
#34. What is the statistical hypothesis? Explain the errors in hypothesis testing.b) Explain the Sample. What are Large Samples & Small Samples?

Statistical Hypothesis: A testable statement about a population parameter.

Errors in Hypothesis Testing:

Type I Error: Rejecting a true null hypothesis.
Type II Error: Failing to reject a false null hypothesis.
Sample: A subset of a population used for analysis.

Large Samples:
𝑛
≥
30
n≥30; more reliable. Small Samples:
𝑛
<
30
n<30; less reliable, use t-distribution.

In [ ]:
#35.A random sample of size 25 from a population gives the sample standard derivation to be 9.0. Test the hypothesis that the population standard derivation is 10.5.
#Hint(Use chi-square distribution).

Given:

Sample size = 25, Sample std dev = 9.0, Population std dev = 10.5
Hypotheses:

𝐻
0
H
0
​
 :
𝜎
=
10.5
σ=10.5
𝐻
1
H
1
​
 :
𝜎
≠
10.5
σ

=10.5
Test Statistic:

𝜒
2
=
(
25
−
1
)
×
9
2
10.
5
2
≈
17.63
χ
2
 =
10.5
2

(25−1)×9
2

​
 ≈17.63
Conclusion:

Compare
𝜒
2
χ
2
  (17.63) with chi-square table value for 24 df. If
𝜒
2
χ
2
  is outside the critical range, reject
𝐻
0
H
0
​
 ; otherwise, do not reject
𝐻
0
H
0
​
 .

In [ ]:
#37.100 students of a PW IOI obtained the following grades in Data Science paper: Grade :[A, B, C, D, E]
#Total Frequency:[15, 17, 30, 22, 16, 100]
#Using the x 2 test, examine the hypothesis that the distribution of grades is uniform.

Given:

Observed Frequencies: [15, 17, 30, 22, 16]
Total = 100 students, Expected Frequency = 20 per grade
Hypotheses:

𝐻
0
H
0
​
 : Uniform distribution
𝐻
1
H
1
​
 : Not uniform
Chi-square Statistic:

𝜒
2
=
(
15
−
20
)
2
20
+
(
17
−
20
)
2
20
+
(
30
−
20
)
2
20
+
(
22
−
20
)
2
20
+
(
16
−
20
)
2
20
=
6.75
χ
2
 =
20
(15−20)
2

​
 +
20
(17−20)
2

​
 +
20
(30−20)
2

​
 +
20
(22−20)
2

​
 +
20
(16−20)
2

​
 =6.75
Degrees of Freedom:
𝑑
𝑓
=
4
df=4

Conclusion:

Compare
𝜒
2
χ
2
  (6.75) with critical value at
𝑑
𝑓
=
4
df=4. Reject
𝐻
0
H
0
​
  if
𝜒
2
χ
2
  is greater than the critical value; otherwise, do not reject
𝐻
0
H
0
​
 .








In [ ]:
#38.Anova Test:
To study the performance of three detergents and three different water temperatures the following whiteness readings were obtained with specially designed equipment.
Water temp            Detergents A        Detergents B        Detergents C
Cold Water                57                    55                67
Worm Water                49                    52                68
Hot Water                 54                    46                58

Calculate Group Means:

Cold Water Mean:
57
+
55
+
67
3
=
59.67
3
57+55+67
​
 =59.67
Warm Water Mean:
49
+
52
+
68
3
=
56.33
3
49+52+68
​
 =56.33
Hot Water Mean:
54
+
46
+
58
3
=
52.67
3
54+46+58
​
 =52.67
Calculate Overall Mean:

Overall Mean:
57
+
55
+
67
+
49
+
52
+
68
+
54
+
46
+
58
9
=
56.33
9
57+55+67+49+52+68+54+46+58
​
 =56.33
Compute ANOVA Table:

Between-Groups Variance (due to differences in water temperature)
Within-Groups Variance (due to differences within each water temperature)
Compare F-Statistic:

Calculate the F-statistic and compare it with the critical value from the F-distribution table with appropriate degrees of freedom.
Conclusion:

If the F-statistic is greater than the critical value, there are significant differences between the groups; otherwise, there are no significant differences.








In [ ]:
#39.How would you create a basic Flask route that displays "Hello, World!" on the homepage?

In [ ]:
from flask import Flask

app = Flask(__name__)

@app.route('/')
def home():
    return 'Hello, World!'

if __name__ == '__main__':
    app.run(debug=True)


 * Serving Flask app '__main__'
 * Debug mode: on


INFO:werkzeug:WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:5000
INFO:werkzeug:Press CTRL+C to quit
INFO:werkzeug: * Restarting with stat


In [ ]:
#40.Explain how to set up a Flask application to handle form submissions using POST requests.

In [ ]:
from flask import Flask, request, render_template

app = Flask(__name__)

@app.route('/', methods=['GET', 'POST'])
def handle_form():
    if request.method == 'POST':
        data = request.form['input_field_name']  # Replace with your form field name
        return f'Form submitted with data: {data}'
    return render_template('form.html')

if __name__ == '__main__':
    app.run(debug=True)


In [ ]:
#41.Write a Flask route that accepts a parameter in the URL and displays it on the page.

In [ ]:
from flask import Flask

app = Flask(__name__)

@app.route('/<name>')
def greet(name):
    return f'Hello, {name}!'

if __name__ == '__main__':
    app.run(debug=True)


In [ ]:
#42.How can you implement user authentication in a Flask application?

In [ ]:
from flask import Flask, render_template_string, redirect, request
from flask_login import LoginManager, UserMixin, login_user, login_required, logout_user, current_user

app = Flask(__name__)
app.secret_key = 'secret_key'
login_manager = LoginManager()
login_manager.init_app(app)

class User(UserMixin):
    def __init__(self, id):
        self.id = id

users = {'user': 'password'}

@login_manager.user_loader
def load_user(user_id):
    return User(user_id)

@app.route('/login', methods=['POST', 'GET'])
def login():
    if request.method == 'POST':
        username = request.form['username']
        password = request.form['password']
        if users.get(username) == password:
            user = User(username)
            login_user(user)
            return redirect('/profile')
    return render_template_string('''
        <form method="post">
            <input type="text" name="username" placeholder="Username">
            <input type="password" name="password" placeholder="Password">
            <input type="submit" value="Login">
        </form>
    ''')

@app.route('/profile')
@login_required
def profile():
    return f'Hello, {current_user.id}!'

@app.route('/logout')
def logout():
    logout_user()
    return redirect('/login')

if __name__ == '__main__':
    app.run(debug=True)


In [ ]:
#43.Describe the process of connecting a Flask app to a SQLite database using SQLAlchemy.

In [ ]:
from flask import Flask
from flask_sqlalchemy import SQLAlchemy

app = Flask(__name__)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///example.db'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
db = SQLAlchemy(app)

class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    username = db.Column(db.String(80), unique=True, nullable=False)

@app.route('/')
def index():
    return 'Hello, SQLAlchemy!'

if __name__ == '__main__':
    db.create_all()  # Create database tables
    app.run(debug=True)


In [ ]:
#44.How would you create a RESTful API endpoint in Flask that returns JSON data?

In [ ]:
from flask import Flask, jsonify

app = Flask(__name__)

@app.route('/api/data')
def get_data():
    return jsonify({"message": "Hello, World!", "status": "success"})

if __name__ == '__main__':
    app.run(debug=True)


In [ ]:
#45.Explain how to use Flask-WTF to create and validate forms in a Flask application.

In [ ]:
from flask import Flask, render_template_string, redirect, url_for
from flask_wtf import FlaskForm
from wtforms import StringField, SubmitField
from wtforms.validators import DataRequired

app = Flask(__name__)
app.secret_key = 'secret_key'

class MyForm(FlaskForm):
    name = StringField('Name', validators=[DataRequired()])
    submit = SubmitField('Submit')

@app.route('/', methods=['GET', 'POST'])
def index():
    form = MyForm()
    if form.validate_on_submit():
        return redirect(url_for('success'))
    return render_template_string('''
        <form method="post">
            {{ form.hidden_tag() }}
            {{ form.name.label }} {{ form.name() }}
            {{ form.submit() }}
        </form>
    ''', form=form)

@app.route('/success')
def success():
    return 'Form successfully submitted!'

if __name__ == '__main__':
    app.run(debug=True)


In [ ]:
#46.How can you implement file uploads in a Flask application?

In [ ]:
from flask import Flask, request, redirect, url_for
import os

app = Flask(__name__)
app.config['UPLOAD_FOLDER'] = 'uploads'
app.config['MAX_CONTENT_LENGTH'] = 16 * 1024 * 1024  # 16 MB limit

if not os.path.exists(app.config['UPLOAD_FOLDER']):
    os.makedirs(app.config['UPLOAD_FOLDER'])

@app.route('/', methods=['GET', 'POST'])
def upload_file():
    if request.method == 'POST':
        if 'file' not in request.files:
            return 'No file part'
        file = request.files['file']
        if file.filename == '':
            return 'No selected file'
        file.save(os.path.join(app.config['UPLOAD_FOLDER'], file.filename))
        return redirect(url_for('upload_file'))
    return '''
    <!doctype html>
    <title>Upload File</title>
    <h1>Upload File</h1>
    <form method=post enctype=multipart/form-data>
      <input type=file name=file>
      <input type=submit value=Upload>
    </form>
    '''

if __name__ == '__main__':
    app.run(debug=True)


In [ ]:
#47.Describe the steps to create a Flask blueprint and why you might use one.

In [ ]:
from flask import Flask, Blueprint, render_template_string

# Create a Blueprint
my_blueprint = Blueprint('my_blueprint', __name__)

@my_blueprint.route('/hello')
def hello():
    return 'Hello from Blueprint!'

# Create the Flask app and register the Blueprint
app = Flask(__name__)
app.register_blueprint(my_blueprint, url_prefix='/blueprint')

@app.route('/')
def index():
    return 'Hello from Main App!'

if __name__ == '__main__':
    app.run(debug=True)


In [ ]:
#48.How would you deploy a Flask application to a production server using Gunicorn and Nginx?

In [ ]:
# 1. Install Gunicorn
pip install gunicorn

# 2. Run Gunicorn (replace `app:app` with your Flask app's module and instance)
gunicorn -w 4 -b 0.0.0.0:8000 app:app

# 3. Configure Nginx (create `/etc/nginx/sites-available/your_app` with the following content)
cat <<EOL | sudo tee /etc/nginx/sites-available/your_app
server {
    listen 80;
    server_name your_domain.com;

    location / {
        proxy_pass http://127.0.0.1:8000;
        proxy_set_header Host \$host;
        proxy_set_header X-Real-IP \$remote_addr;
        proxy_set_header X-Forwarded-For \$proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto \$scheme;
    }
}
EOL

# 4. Enable Nginx Configuration
sudo ln -s /etc/nginx/sites-available/your_app /etc/nginx/sites-enabled/
sudo systemctl restart nginx


In [ ]:
#49. Make a fully functional web application using flask, Mangodb. Signup,Signin page.And after successfully login .Say hello Geeks message at webpage.

In [ ]:
from flask import Flask, render_template_string, request, redirect, url_for, session
from flask_pymongo import PyMongo
from werkzeug.security import generate_password_hash, check_password_hash

app = Flask(__name__)
app.secret_key = 'secret_key'
app.config['MONGO_URI'] = 'mongodb://localhost:27017/mydatabase'
mongo = PyMongo(app)

# Signup and Signin HTML
signup_html = '''
<form method="post" action="/signup">
    <input type="text" name="username" placeholder="Username" required>
    <input type="password" name="password" placeholder="Password" required>
    <input type="submit" value="Signup">
</form>
<a href="/signin">Signin</a>
'''

signin_html = '''
<form method="post" action="/signin">
    <input type="text" name="username" placeholder="Username" required>
    <input type="password" name="password" placeholder="Password" required>
    <input type="submit" value="Signin">
</form>
<a href="/signup">Signup</a>
'''

@app.route('/signup', methods=['GET', 'POST'])
def signup():
    if request.method == 'POST':
        username = request.form['username']
        password = generate_password_hash(request.form['password'])
        mongo.db.users.insert_one({'username': username, 'password': password})
        return redirect(url_for('signin'))
    return render_template_string(signup_html)

@app.route('/signin', methods=['GET', 'POST'])
def signin():
    if request.method == 'POST':
        username = request.form['username']
        password = request.form['password']
        user = mongo.db.users.find_one({'username': username})
        if user and check_password_hash(user['password'], password):
            session['username'] = username
            return redirect(url_for('hello'))
    return render_template_string(signin_html)

@app.route('/hello')
def hello():
    if 'username' in session:
        return f'Hello, {session["username"]}!'
    return redirect(url_for('signin'))

if __name__ == '__main__':
    app.run(debug=True)
